src/org/sosy_lab/java_smt/api/Model.java

```diff
@@ -52,11 +52,15 @@ public interface Model extends Evaluator, Iterable<ValueAssignment>, AutoCloseab
    */
   @Override
   default Iterator<ValueAssignment> iterator() {
-    return asList().iterator();
+    try {
+      return asList().iterator();
+    } catch (InterruptedException | SolverException e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
I think this should be improved.

First, it is absolutely clear that this needs proper documentation that tells users what to expect and that they should not use it if if they care about handling these exceptions. This should be added to the JavaDoc of the method as well as of the class and maybe even to `getModel()`, because users are unlikely to read the JavaDoc of an `iterator()` method that can be called implicitly in a for-each loop.

Second, using a `RuntimeException` is very generic and makes it look like a bug if it is thrown, even if it is not. Users who want to catch and handle these exceptions need to catch `RuntimeException` and risk catching too many exceptions. The API would be much nicer if a specific exception type was used that signals that some expected checked exception is contained that was not possible to be thrown directly. Maybe there even exists such an exception type already somewhere.

Third, for catching `InterruptedException` I think that the code should re-add the interrupted status of the thread.

In general it would also probably be a good idea to look at how other code with the same problem is designed and handles checked exceptions. AFAIR there are for example cases in the Java API where streams are used for I/O and checked exceptions are a problem, and there are probably several other such cases in the Java API and maybe Guava. If we look at their solutions, we do have to invest less time into finding our own good solution.

Apart from the API change here, I just noticed that if I did not overlook anything, `Model.iterator()` seems to be never overwritten in any of the solver-specific classes. Why is that?

I was under the impression as a JavaSMT user that I have two ways for iterating over model values: Either I retrieve all model values into a list with `asList()` or `getModelAssignments()` and have to pay for the initial creation and memory consumption, but can then cheaply iterate over all values repeatedly. Or I could iterate using `.iterator()` and retrieve values lazily from the solver, without O(n) memory consumption and upfront cost. The JavaDocs of `getModelAssignments()` and `iterator()` even still indicate that this performance difference may exist.

I wonder why is this no longer possible? Could this be re-added?

If the answer to the last question is "no", then the question remains why JavaSMT even offers two ways to retrieve model values that are in all circumstances the same. Every caller could easily use `getModel().asList().iterator()` or `getModelAssignments().iterator()` instead of `getModel().iterator`, this is not too much burden. In this case the `iterator()` method and the inheritance from `Iterable` could simply be marked as `@Deprecated` with the exception problem and a reference to `asList()` as explanation.

But in general I would prefer if a proper implementation of lazy model iteration is re-added.